### PR TITLE
Airlock API (Draft + Submit)

### PR DESCRIPTION
**Pull request by anatbal**
# PR for issue #1832 

## What is being addressed

First draft of the API endpoints for airlock submit and request.

### CODE REVIEW

File: api_app/models/schemas/airlock_request.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    return {
+        "requestId": airlock_request_id,
+        "workspaceId": workspace_id,
+        "isActive": True,
+        "requestType": "import",
+        "files": [],
+        "business_justification": "some business justification",
+        "status": "draft",
+        "properties": {
+            "display_name": "my workspace service",
```
Comment (eladiw) [inline api_app/models/schemas/airlock_request.py:16]:
```suggestion
            "display_name": "some property",
```

File: api_app/api/routes/workspaces.py [hunk 2]
```diff
@@ -384,6 +387,17 @@ async def retrieve_user_resource_operations_by_user_resource_id(user_resource=De
 
 
 @user_resources_workspace_router.get("/workspaces/{workspace_id}/workspace-services/{service_id}/user-resources/{resource_id}/operations/{operation_id}", response_model=OperationInResponse, name=strings.API_GET_RESOURCE_OPERATION_BY_ID, dependencies=[Depends(get_workspace_by_id_from_path)])
-async def retrieve_user_resource_operations_by_user_resource_id_and_operation_id(user_resource=Depends(get_user_resource_by_id_from_path), user=Depends(get_current_workspace_owner_or_researcher_user), operation=Depends(get_operation_by_id_from_path)) -> OperationInList:
+async def retrieve_user_resosurce_operations_by_user_resource_id_and_operation_id(user_resource=Depends(get_user_resource_by_id_from_path), user=Depends(get_current_workspace_owner_or_researcher_user), operation=Depends(get_operation_by_id_from_path)) -> OperationInList:
```
Comment (ross-p-smith) [inline api_app/api/routes/workspaces.py:390]:
```suggestion
async def retrieve_user_resource_operations_by_user_resource_id_and_operation_id(user_resource=Depends(get_user_resource_by_id_from_path), user=Depends(get_current_workspace_owner_or_researcher_user), operation=Depends(get_operation_by_id_from_path)) -> OperationInList:
```

File: api_app/models/domain/airlock_resource.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
+from enum import Enum
+from typing import List
+from pydantic import Field
+from models.domain.azuretremodel import AzureTREModel
+from resources import strings
+
+
+class AirlockResourceType(str, Enum):
+    """
+    Type of resource to deploy
```
Comment (eladiw) [inline api_app/models/domain/airlock_resource.py:10]:
```suggestion
    Type of resource to create
```

File: api_app/models/domain/airlock_resource.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    """
+    AirlockRequest = strings.AIRLOCK_RESOURCE_TYPE_REQUEST
+    # TBD Airlock review
+
+
+class AirlockResourceHistoryItem(AzureTREModel):
```
Comment (eladiw) [inline api_app/models/domain/airlock_resource.py:16]: are we going to implement history?

File: api_app/models/domain/airlock_resource.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    resourceVersion: int
+    updatedWhen: float
+    user: dict = {}
+
+
+class AirlockResource(AzureTREModel):
+    """
+    Resource request
+    """
+    id: str = Field(title="Id", description="GUID identifying the resource request")
```
Comment (eladiw) [inline api_app/models/domain/airlock_resource.py:31]:
```suggestion
    id: str = Field(title="Id", description="GUID identifying the resource")
```

File: api_app/models/domain/airlock_request.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    """
+    Airlock request
+    """
+    workspaceId: str = Field("", title="Workspace ID", description="Service target Workspace id")
+    resourceType = AirlockResourceType.AirlockRequest
+    requestType: AirlockRequestType = Field("", title="Airlock request type")
+    files: List[str] = Field([], title="Files of the request")
+    business_justification: str = Field("Business Justifications", title="Explanation that will be provided to the request reviewer")
+    status: AirlockRequestStatus = Field(AirlockRequestStatus.Draft, title="Airlock request status")
+    container: List[AirlockContainer] = Field([], title="Containers of the request")
```
Comment (eladiw) [inline api_app/models/domain/airlock_request.py:42]: I think the containers list should not be part of the Request schema.
can we add this just for the RequestInResponse?wdyt

File: api_app/api/routes/workspaces.py [hunk 7]
```diff
@@ -12,12 +12,14 @@
 from db.repositories.user_resources import UserResourceRepository
 from db.repositories.workspaces import WorkspaceRepository
 from db.repositories.workspace_services import WorkspaceServiceRepository
+from db.repositories.airlock_requets import AirlockRequestRepository
```
Comment (damoodamoo) [inline api_app/api/routes/workspaces.py:15]: `nit` - typo here - `requests`

File: api_app/models/domain/airlock_resource.py (hunk 4 above)
Comment (anatbal) [reply api_app/models/domain/airlock_resource.py:16]: If we follow the convention, in the "base" resource, history is implemented there

File: api_app/models/domain/airlock_request.py (hunk 6 above)
Comment (anatbal) [reply api_app/models/domain/airlock_request.py]: Was removed

File: api_app/event_grid/helpers.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+from azure.eventgrid.aio import EventGridPublisherClient
+from azure.core.credentials import AzureKeyCredential
+from models.domain.airlock_request import AirlockRequest
+from core import config
+
+
+async def _publish_event(event: EventGridEvent, topic_key: str, topic_endpoint: str):
+    credential = AzureKeyCredential(topic_key)
+    client = EventGridPublisherClient(topic_endpoint, credential)
+
```
Comment (eladiw) [inline api_app/event_grid/helpers.py:12]: add logs

File: api_app/event_grid/helpers.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    status_changed_event = EventGridEvent(
+        event_type="statusChanged",
+        data={
+            "request_id": request_id,
+            "status": status,
+            "type": request_type,
+            "workspace_id": workspace_id
+        },
+        subject=f"{request_id}/statusChanged",
```
Comment (eladiw) [inline api_app/event_grid/helpers.py:30]: check if this is the right value for the subject

**Review by eladiw — CHANGES_REQUESTED** [review]
🚀

File: templates/core/terraform/api-webapp.tf [hunk 10]
```diff
@@ -37,6 +53,8 @@ resource "azurerm_app_service" "api" {
     "STATE_STORE_ENDPOINT"                       = azurerm_cosmosdb_account.tre-db-account.endpoint
     "COSMOSDB_ACCOUNT_NAME"                      = azurerm_cosmosdb_account.tre-db-account.name
     "SERVICE_BUS_FULLY_QUALIFIED_NAMESPACE"      = "sb-${var.tre_id}.servicebus.windows.net"
+    "EVENT_GRID_TOPIC_ENDPOINT"                  =  data.azurerm_eventgrid_topic.status_changed.endpoint
```
[the remark this answers is not in the document]
Comment (anatbal) [reply templates/core/terraform/api-webapp.tf]: Fixed

File: api_app/api/routes/airlock_resource_helpers.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+
+async def save_and_publish_event_airlock_request(resource: AirlockRequest, resource_repo: AirlockRequestRepository, user: User):
+    try:
+        resource.user = user
+        resource.updatedWhen = get_timestamp()
+        resource_repo.save_item(resource)
+    except Exception as e:
+        logging.error(f'Failed saving resource item {resource}: {e}')
+        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=strings.STATE_STORE_ENDPOINT_NOT_RESPONDING)
```
[the remark this answers is not in the document]
Comment (anatbal) [reply api_app/api/routes/airlock_resource_helpers.py:21]: It's the API convention, used in many other places. I am not sure if it's the best, but it already being used so would open a separate issue if we want to change it

File: api_app/api/routes/airlock_resource_helpers.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+    try:
+        await send_status_changed_event(resource)
+    except Exception as e:
+        resource_repo.delete_item(resource.id)
+        logging.error(f"Failed send resource request message: {e}")
+        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=strings.EVENT_GRID_GENERAL_ERROR_MESSAGE)
```
[the remark this answers is not in the document]
Comment (anatbal) [reply api_app/api/routes/airlock_resource_helpers.py:28]: same as above

File: api_app/api/routes/workspaces.py [hunk 13]
```diff
@@ -6,18 +6,22 @@
 
 from api.dependencies.database import get_repository
```
Comment (damoodamoo) [inline api_app/api/routes/workspaces.py:7]: Can we avoid putting more into this one python file? With airlock it feels like a good candidate to be in it's own file?

File: api_app/api/routes/airlock_resource_helpers.py [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+from starlette import status
+from db.repositories.airlock_requests import AirlockRequestRepository
+from models.domain.airlock_request import AirlockRequest, AirlockRequestStatus
+from event_grid.helpers import send_status_changed_event
+from models.domain.authentication import User
+
+from resources import strings
+
+
+async def save_and_publish_event_airlock_request(resource: AirlockRequest, resource_repo: AirlockRequestRepository, user: User):
```
Comment (damoodamoo) [inline api_app/api/routes/airlock_resource_helpers.py:14]: can we change `resource_repo` to `airlock_repo` to avoid confusion?

**Review by damoodamoo — COMMENTED** [review]
Nice work - have a few comments but nothing big - I won't 'approve' as someone who understands what airlock is doing should do that :) 

Also are you planning to add tests in a separate task?

File: templates/core/terraform/api-webapp.tf [hunk 15]
```diff
@@ -2,6 +2,22 @@ data "local_file" "api_app_version" {
   filename = "${path.root}/../../../api_app/_version.py"
 }
 
+data "azurerm_eventgrid_topic" "status_changed" {
+  name                = "evgt-airlock-status-changed-${var.tre_id}"
+  resource_group_name = azurerm_resource_group.core.name
+    depends_on = [
+    module.airlock_resources
+  ]
```
[the remark this answers is not in the document]
Comment (eladiw) [reply templates/core/terraform/api-webapp.tf]: talked offline

File: api_app/db/repositories/airlock_resources.py [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+        super().__init__(client, config.STATE_STORE_AIRLOCK_RESOURCES_CONTAINER)
+
+    @staticmethod
+    def get_resource_base_spec_params():
+        return {"tre_id": config.TRE_ID}
+
+    def get_timestamp(self) -> float:
+        return datetime.utcnow().timestamp()
+
+    def update_airlock_resource_item(self, original_resource: AirlockResource, new_resource: AirlockResource, user: User) -> AirlockResource:
```
[the remark this answers is not in the document]
Comment (anatbal) [reply api_app/db/repositories/airlock_resources.py:20]: I added that logic in a different place, under airlock_requests, let me know wdyt

File: api_app/models/domain/airlock_resource.py [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+    # TBD Airlock review
+
+
+class AirlockResourceHistoryItem(AzureTREModel):
+    """
+    Resource History Item - to preserve history of resource properties
+    """
+    resourceVersion: int
+    updatedWhen: float
+    user: dict = {}
```
[the remark this answers is not in the document]
Comment (anatbal) [reply api_app/models/domain/airlock_resource.py:35]: Added the status change

File: templates/core/terraform/api-webapp.tf [hunk 18]
```diff
@@ -37,6 +53,8 @@ resource "azurerm_app_service" "api" {
     "STATE_STORE_ENDPOINT"                       = azurerm_cosmosdb_account.tre-db-account.endpoint
     "COSMOSDB_ACCOUNT_NAME"                      = azurerm_cosmosdb_account.tre-db-account.name
     "SERVICE_BUS_FULLY_QUALIFIED_NAMESPACE"      = "sb-${var.tre_id}.servicebus.windows.net"
+    "EVENT_GRID_TOPIC_ENDPOINT"                  = data.azurerm_eventgrid_topic.status_changed.endpoint
+    "EVENT_GRID_ACCESS_KEY"                      = "@Microsoft.KeyVault(SecretUri=${data.azurerm_key_vault_secret.eventgrid_status_changed_access_key.id})"
```
Comment (tamirkamara) [inline templates/core/terraform/api-webapp.tf:57]: Now I see... I prefer this won't be here and we use the process managed identity to access the resource.

File: api_app/db/repositories/airlock_resources.py [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+        )
+        new_resource.history.append(history_item)
+
+        # now update the resource props
+        new_resource.resourceVersion = new_resource.resourceVersion + 1
+        new_resource.user = user
+        new_resource.updatedWhen = self.get_timestamp()
+
+        # TODO add validation
+        self.update_item(new_resource)
```
[the remark this answers is not in the document]
Comment (anatbal) [reply api_app/db/repositories/airlock_resources.py:35]: The log lines already appear whenever this fun is being refrenced

File: devops/scripts/control_tre.sh [hunk 20]
```diff
@@ -16,7 +16,7 @@ if [[ $(az group list --output json --query "[?name=='rg-${TRE_ID}'] | length(@)
   exit 0
 fi
 
-az config set extension.use_dynamic_install=yes_without_prompt
+sudo az config set extension.use_dynamic_install=yes_without_prompt
```
[the remark this answers is not in the document]
Comment (anatbal) [reply devops/scripts/control_tre.sh]: Removed

File: api_app/db/repositories/airlock_requests.py [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+        super().__init__(client)
+
+    @staticmethod
+    def airlock_request_query_string():
+        return f'SELECT * FROM c WHERE c.resourceType = "{AirlockResourceType.AirlockRequest}"'
+
+    def create_airlock_request_item(self, airlock_request_input: AirlockRequestInCreate, workspace_id: str) -> AirlockRequest:
+        full_airlock_request_id = str(uuid.uuid4())
+
+        # TODO - validate
```
[the remark this answers is not in the document]
Comment (anatbal) [reply api_app/db/repositories/airlock_requests.py]: Added

File: api_app/models/domain/airlock_request.py [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+
+class AirlockRequest(AirlockResource):
+    """
+    Airlock request
+    """
+    workspaceId: str = Field("", title="Workspace ID", description="Service target Workspace id")
+    resourceType = AirlockResourceType.AirlockRequest
+    requestType: AirlockRequestType = Field("", title="Airlock request type")
+    files: List[str] = Field([], title="Files of the request")
+    business_justification: str = Field("Business Justifications", title="Explanation that will be provided to the request reviewer")
```
[the remark this answers is not in the document]
Comment (anatbal) [reply api_app/models/domain/airlock_request.py]: Agree, changed

File: api_app/db/repositories/airlock_resources.py [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+            user=original_resource.user
+        )
+        new_resource.history.append(history_item)
+
+        # now update the resource props
+        new_resource.resourceVersion = new_resource.resourceVersion + 1
+        new_resource.user = user
+        new_resource.updatedWhen = self.get_timestamp()
+
+        # TODO add validation
```
[the remark this answers is not in the document]
Comment (anatbal) [reply api_app/db/repositories/airlock_resources.py]: added

File: api_app/models/domain/airlock_resource.py [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+from models.domain.azuretremodel import AzureTREModel
+from resources import strings
+
+
+class AirlockResourceType(str, Enum):
+    """
+    Type of resource to create
+    """
+    AirlockRequest = strings.AIRLOCK_RESOURCE_TYPE_REQUEST
+    # TBD Airlock review
```
[the remark this answers is not in the document]
Comment (anatbal) [reply api_app/models/domain/airlock_resource.py:26]: Done

File: api_app/api/routes/workspaces.py (hunk 13 above)
Comment (anatbal) [reply api_app/api/routes/workspaces.py]: Moved.

**Comment by anatbal** [comment]
@damoodamoo I already opened a separate issue for testing and validation - https://github.com/microsoft/AzureTRE/issues/2016

File: api_app/event_grid/helpers.py (hunk 8 above)
Comment (anatbal) [reply api_app/event_grid/helpers.py:12]: The logging happens is the function that calls that one

File: templates/core/terraform/api-webapp.tf (hunk 18 above)
Comment (anatbal) [reply templates/core/terraform/api-webapp.tf:57]: https://github.com/microsoft/AzureTRE/issues/2019

**Comment by anatbal** [comment]
test-extended

File: api_app/db/repositories/airlock_requests.py [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+        # If draft can only be changed to submitted
+        draft_condition = current_status == AirlockRequestStatus.Draft and new_status == AirlockRequestStatus.Submitted
+        # If submitted needs to get a review first
+        submit_condition = current_status == AirlockRequestStatus.Submitted and new_status == AirlockRequestStatus.InReview
+        # If in review can only be changed to either approve or rejected
+        in_review_condition = current_status == AirlockRequestStatus.InReview and new_status == AirlockRequestStatus.Approved or new_status == AirlockRequestStatus.Rejected
+        # If rejected can only return to in review or be blocked
+        rejected_condition = current_status == AirlockRequestStatus.Rejected and new_status == AirlockRequestStatus.InReview or new_status == AirlockRequestStatus.Blocked
+
+        return cancel_condition and blocked_condition and approved_condition and draft_condition and submit_condition and in_review_condition and rejected_condition
```
Comment (eladiw) [inline api_app/db/repositories/airlock_requests.py:36]: nice

File: api_app/models/schemas/airlock_request.py [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+    class Config:
+        schema_extra = {
+            "example": {
+                "airlock_request": get_sample_airlock_request("933ad738-7265-4b5f-9eae-a1a62928772e", "121e921f-a4aa-44b3-90a9-e8da030495ef")
+            }
+        }
+
+
+class AirlockRequestInCreate(BaseModel):
+    requestType: AirlockRequestType = Field("", title="Airlock request type", description="Values for the parameters required by the Airlock request specification")
```
[the remark this answers is not in the document]
Comment (eladiw) [reply api_app/models/schemas/airlock_request.py:30]: marked as resolved but didn't approve or commented why not...

**Review by eladiw — APPROVED** [review]
lgtm. minor comments

File: api_app/db/repositories/airlock_requests.py [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+        blocked_condition = current_status != AirlockRequestStatus.Blocked
+        # Cannot change status from approved
+        approved_condition = current_status != AirlockRequestStatus.Approved
+        # If draft can only be changed to submitted
+        draft_condition = current_status == AirlockRequestStatus.Draft and new_status == AirlockRequestStatus.Submitted
+        # If submitted needs to get a review first
+        submit_condition = current_status == AirlockRequestStatus.Submitted and new_status == AirlockRequestStatus.InReview
+        # If in review can only be changed to either approve or rejected
+        in_review_condition = current_status == AirlockRequestStatus.InReview and new_status == AirlockRequestStatus.Approved or new_status == AirlockRequestStatus.Rejected
+        # If rejected can only return to in review or be blocked
```
[the remark this answers is not in the document]
Comment (anatbal) [reply api_app/db/repositories/airlock_requests.py]: fixed

**Review by eladiw — APPROVED** [review]
lgtm. minor comments

File: templates/core/terraform/airlock/eventgrid_topics.tf [hunk 28]
```diff
@@ -19,6 +24,15 @@ resource "azurerm_eventgrid_topic" "status_changed" {
   }
 }
 
+resource "azurerm_key_vault_secret" "eventgrid_status_changed_access_key" {
```
[the remark this answers is not in the document]
Comment (anatbal) [reply templates/core/terraform/airlock/eventgrid_topics.tf:64]: Issue was opened as discussed #2019

**Pull request merged**